Initial commit. Basic thermometer for one participant working

adinaelisheva opened the pull request:
lots of todos as noted, especially if we're going to be publishing this in any sort of public forum. let me know anything else that could be changed.

adinaelisheva committed:
Initial commit. Basic thermometer for one participant working

karantza commented on Widgets/widget.css:
```diff
@@ -0,0 +1,71 @@
+div#achieved {
+  position: absolute;
+  top: 5px;
+  left: calc(50% - 105px);
```
I wonder if we should make a policy of commenting whenever we use calc, since I have no idea where these numbers come from

adinaelisheva replied:
good idea, oops

karantza commented on js/ParticipantProgress.js:
```diff
@@ -1,7 +1,135 @@
+
+(function() {
+  //copied from http://papermashup.com/read-url-get-variables-withjavascript/
+  function getUrlVars() {
```
can we put this in the globals.js file so we only have one copy of it?

adinaelisheva replied:
yeah good call, i'll add it

karantza commented on js/ParticipantProgress.js:
```diff
@@ -1,7 +1,135 @@
+    var uid = urlVars['id'];
+
+    //borrowed from http://axonflux.com/handy-rgb-to-hsl-and-rgb-to-hsv-color-model-c
+    var hslToRgb = function(h, s, l){
```
this could also go in globals

adinaelisheva replied:
is this actually going to be used by anyone else though?

Ganon11 replied:
depends on how fancy the other widgets are. You could add a... I don't know, Colors.js file to do color-related functions.

adinaelisheva replied:
i'm gonna wait and see if anyone else actually needs it before doing that. but i'll keep it in mind

karantza commented on js/ParticipantProgress.js:
```diff
@@ -1,7 +1,135 @@
+        "background-color": color
+      };
+      //hacky text measurement - add 60 for 'total: $'
+      var textWidth = 60 + 13 * $scope.curTotal.toString().length;
```
I really wish there was some way to get these values from the DOM rather than hardcoding them, but I have no idea if that is possible :(

adinaelisheva replied:
there is! it involves making a hidden div with the text `"total: $<blah>"` in it, measuring it, and then deleting it. but i didn't feel like it.... i can add it as a todo. you can put it in a fn so it isn't too bad. http://stackoverflow.com/questions/118241/calculate-text-width-with-javascript

karantza commented on js/ParticipantProgress.js:
```diff
@@ -1,7 +1,135 @@
+    
+    //TODO - better error handling. probably a uniform error page btw all widgets
+    var error = function(){
+      console.log('Bad user ID in URL');
```
I experimented with this, and a bad ID doesn't actually trigger an error, at least not for donation_info. The errors will probably be if, like, extralife goes down... maybe make the error functions inline with the calls on lines 126/127, so they each have different messages? Not sure if that's good style. Also it doesn't really matter until/if we display them to the user...

adinaelisheva replied:
I checked and it looks like it's erroring twice (whether this is desired, i'm not sure...)
The reason i'm doing it this way is because the only info they're giving us is the user ID, so I figured that's the only thing that they could be doing wrong. If the server's down, I don't know how to tell, and also what can they do about it?
I did comment saying i want to add better error handling. My vague plan is to abort out if the first command failed, and as I mentioned, make an actual error page. That error page could say something like "Fetching data for user <num> failed. Is the number correct, or is the server down?"

karantza commented on js/ParticipantProgress.js:
```diff
@@ -1,7 +1,135 @@
+
+  });
+
+})();
```
This is probably good practice...

Ganon11 commented on js/ParticipantProgress.js:
```diff
@@ -1,7 +1,135 @@
+//     //console.log(personalData);
+//     var donationData = GLOBALS.GET_PARTICIPANT_DONATION_INFO(participantId);
+//     //console.log(donationData);
+// }
```
Just delete this function, I was only using it to mess around with stuff.

adinaelisheva replied:
oops, will do

adinaelisheva commented:
Changes based on people's comments:
- added comments on uses of calc
- deleted mike's old code
- added a TODO to explore actual text measurement
- moved getUrlVars to GLOBALS

I'm gonna merge this now but feel free to comment on this PR posthumously if you still have stuff to say/complain about
